Issue with @JsonbCreator constructor with Generic Type in Yasson 3.0.0 and higher

There seems to be a bug in Yasson version 3.0.0 and higher that affects the `@JsonbCreator` constructor when it has a Generic Type parameter. This issue is not present in Yasson version 2.0.4.


Stack trace from the reproducer (version 3.0.0 and higher):
```
Exception in thread "main" jakarta.json.bind.JsonbException: Error resolving runtime type for type: T
	at org.eclipse.yasson.internal.ReflectionUtils.lambda$getRawType$0(ReflectionUtils.java:92)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at org.eclipse.yasson.internal.ReflectionUtils.getRawType(ReflectionUtils.java:92)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.createObjectDeserializer(DeserializationModelCreator.java:236)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.deserializerChainInternal(DeserializationModelCreator.java:193)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.deserializerChain(DeserializationModelCreator.java:135)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.createNewChain(DeserializationModelCreator.java:488)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.typeProcessor(DeserializationModelCreator.java:477)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.typeProcessor(DeserializationModelCreator.java:430)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.createCollectionDeserializer(DeserializationModelCreator.java:273)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.deserializerChainInternal(DeserializationModelCreator.java:183)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.deserializerChain(DeserializationModelCreator.java:135)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.createNewChain(DeserializationModelCreator.java:488)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.typeProcessor(DeserializationModelCreator.java:477)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.typeProcessor(DeserializationModelCreator.java:430)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.createObjectDeserializer(DeserializationModelCreator.java:227)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.deserializerChainInternal(DeserializationModelCreator.java:193)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.deserializerChain(DeserializationModelCreator.java:135)
	at org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.deserializerChain(DeserializationModelCreator.java:123)
	at org.eclipse.yasson.internal.DeserializationContextImpl.deserializeItem(DeserializationContextImpl.java:137)
	at org.eclipse.yasson.internal.DeserializationContextImpl.deserialize(DeserializationContextImpl.java:127)
	at org.eclipse.yasson.internal.JsonBinding.deserialize(JsonBinding.java:55)
	at org.eclipse.yasson.internal.JsonBinding.fromJson(JsonBinding.java:62)
	at Reproducer.main(Reproducer.java:27)
```

**Reproducer using JBang**:
```java
///usr/bin/env jbang "$0" "$@" ; exit $?
//DEPS org.eclipse:yasson:3.0.3


import jakarta.json.bind.Jsonb;
import jakarta.json.bind.JsonbConfig;
import jakarta.json.bind.JsonbBuilder;
import jakarta.json.bind.annotation.JsonbCreator;
import java.util.List;


public class Reproducer {

    public static void main(String... args) {
        Jsonb jsonb = JsonbBuilder.create(new JsonbConfig()
                .withFormatting(true)
                .withNullValues(true));
        String jsonString = "{\n" +
                "    \"id\": 0,\n" +
                "    \"multiLangRecords\": [{\n" +
                "      \"lang\": \"\",\n" +
                "      \"record\": {\n" +
                "        \"description\": \"\"\n" +
                "      }\n" +
                "    }]  \n" +
                "  }";
        System.out.println(jsonb.fromJson(jsonString, InsertTechnicalLocationInputJ.class));
        // EXPECTED: InsertTechnicalLocationInputJ[id=0, multiLangRecords=[InsertLangEntryGQLJ[lang=, record=InsertTechnicalLocationRecordGQLJ[description=]]]]
    }

    public record InsertTechnicalLocationInputJ(String id,
                                                List<InsertLangEntryGQLJ<InsertTechnicalLocationRecordGQLJ>> multiLangRecords) {

        @JsonbCreator
        public InsertTechnicalLocationInputJ {
        }
    }

    public record InsertTechnicalLocationRecordGQLJ(String description) {

        @JsonbCreator
        public InsertTechnicalLocationRecordGQLJ {
        }
    }

    public record InsertLangEntryGQLJ<T>(String lang, T record) {
        @JsonbCreator
        public InsertLangEntryGQLJ {
        }
    }

}
```
Switching yasson's version (line starting `//DEPS`) from 3.0.3 to 2.0.4 resolves the issue.

JBang documentation and how to download it: https://www.jbang.dev/

It seems we are facing the same issue using Yasson 3.0.2 (which ships with WildFly 30):

```
jakarta.ws.rs.ProcessingException: RESTEASY008200: JSON Binding deserialization error: jakarta.json.bind.JsonbException: Error resolving runtime type for type: T
	at org.jboss.resteasy.resteasy-json-binding-provider@6.2.5.Final//org.jboss.resteasy.plugins.providers.jsonb.JsonBindingProvider.readFrom(JsonBindingProvider.java:78
	
[..]
Caused by: jakarta.json.bind.JsonbException: Error resolving runtime type for type: T
	at org.eclipse.yasson//org.eclipse.yasson.internal.ReflectionUtils.lambda$getRawType$0(ReflectionUtils.java:92)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at org.eclipse.yasson//org.eclipse.yasson.internal.ReflectionUtils.getRawType(ReflectionUtils.java:92)
	at org.eclipse.yasson//org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.createObjectDeserializer(DeserializationModelCreator.java:236)
	at org.eclipse.yasson//org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.deserializerChainInternal(DeserializationModelCreator.java:193)
	at org.eclipse.yasson//org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.deserializerChain(DeserializationModelCreator.java:135)
	at org.eclipse.yasson//org.eclipse.yasson.internal.deserializer.DeserializationModelCreator.deserializerChain(DeserializationModelCreator.java:123)
	at org.eclipse.yasson//org.eclipse.yasson.internal.DeserializationContextImpl.deserializeItem(DeserializationContextImpl.java:137)
	at org.eclipse.yasson//org.eclipse.yasson.internal.DeserializationContextImpl.deserialize(DeserializationContextImpl.java:127)
	at org.eclipse.yasson//org.eclipse.yasson.internal.JsonBinding.deserialize(JsonBinding.java:55)
	at org.eclipse.yasson//org.eclipse.yasson.internal.JsonBinding.fromJson(JsonBinding.java:95)
	at org.jboss.resteasy.resteasy-json-binding-provider@6.2.5.Final//org.jboss.resteasy.plugins.providers.jsonb.ManagedJsonb.fromJson(ManagedJsonb.java:73)
	at org.jboss.resteasy.resteasy-json-binding-provider@6.2.5.Final//org.jboss.resteasy.plugins.providers.jsonb.JsonBindingProvider.readFrom(JsonBindingProvider.java:71)	
```	

In our case the Java data class in question which Yasson is trying to deserialize is:

```
public class RuleResponse<T> {

    private final T result;

    @JsonbCreator
    public RuleResponse(@JsonbProperty("result") final T result) {
        this.result = result;
    }

    public T getResult() {
        return result;
    }
}
```

Is there a workaround for this issue? We cannot just downgrade Yasson because Yasson 3 is used internally by WildFly a lot as well it seems.

--- a/etc/copyright.sh
+++ b/etc/copyright.sh
@@ -16,7 +16,7 @@ die(){ echo "${1}" ; exit 1 ;}
 readonly RESULT_FILE="target/copyright-check.txt"
 mkdir target
 
-mvn -q org.glassfish.copyright:glassfish-copyright-maven-plugin:copyright \
+mvn -q validate -Pcopyright \
         > ${RESULT_FILE} || (cat ${RESULT_FILE}; die "Error running the Maven command")
 
 grep -i "copyright" ${RESULT_FILE} \

--- a/pom.xml
+++ b/pom.xml
@@ -275,6 +275,10 @@
                     <plugin>
                         <groupId>org.glassfish.copyright</groupId>
                         <artifactId>glassfish-copyright-maven-plugin</artifactId>
+                        <configuration>
+                            <!-- Ignore the year to comply with Eclipse Foundation. -->
+                            <ignoreYear>true</ignoreYear>
+                        </configuration>
                         <executions>
                             <!--Copyright goal does not fail when copyright is not properly updated,
                                 but prints out which file has incorrect copyright.-->

--- a/src/main/java/org/eclipse/yasson/internal/ReflectionUtils.java
+++ b/src/main/java/org/eclipse/yasson/internal/ReflectionUtils.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015, 2023 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2015 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0 which is available at
@@ -189,6 +189,10 @@ public class ReflectionUtils {
             Type tmp = new VariableTypeInheritanceSearch().searchParametrizedType(type, (TypeVariable<?>) returnType);
             if (tmp != null) {
                 returnType = tmp;
+            }
+            // If the type is a WildcardType we need to resolve the most specific type
+            if (returnType instanceof WildcardType) {
+                return resolveMostSpecificBound(chain, (WildcardType) returnType, warn);
             }
             if (!(returnType instanceof TypeVariable)) {
                 break;

--- a/src/main/java/org/eclipse/yasson/internal/deserializer/DeserializationModelCreator.java
+++ b/src/main/java/org/eclipse/yasson/internal/deserializer/DeserializationModelCreator.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, 2023 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0 which is available at
@@ -233,7 +233,7 @@ public class DeserializationModelCreator {
             if (creatorModel.getCustomization().isRequired()) {
                 defaultCreatorValues.put(parameterName, new RequiredCreatorParameter(parameterName));
             } else {
-                Class<?> rawParamType = ReflectionUtils.getRawType(creatorModel.getType());
+                Class<?> rawParamType = ReflectionUtils.getOptionalRawType(creatorModel.getType()).orElse(Object.class);
                 defaultCreatorValues.put(parameterName, DEFAULT_CREATOR_VALUES.getOrDefault(rawParamType, NULL_PROVIDER));
             }
         }

--- a/src/test/java/org/eclipse/yasson/defaultmapping/generics/GenericsTest.java
+++ b/src/test/java/org/eclipse/yasson/defaultmapping/generics/GenericsTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015, 2023 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2015 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0 which is available at
@@ -13,6 +13,7 @@
 package org.eclipse.yasson.defaultmapping.generics;
 
 import java.lang.reflect.Type;
+import java.lang.reflect.WildcardType;
 import java.math.BigDecimal;
 import java.text.ParseException;
 import java.text.SimpleDateFormat;
@@ -36,8 +37,11 @@ import org.eclipse.yasson.adapters.model.GenericBox;
 import org.eclipse.yasson.defaultmapping.generics.model.AnotherGenericTestClass;
 import org.eclipse.yasson.defaultmapping.generics.model.BoundedGenericClass;
 import org.eclipse.yasson.defaultmapping.generics.model.Circle;
+import org.eclipse.yasson.defaultmapping.generics.model.CollectionContainer;
+import org.eclipse.yasson.defaultmapping.generics.model.CollectionElement;
 import org.eclipse.yasson.defaultmapping.generics.model.CollectionWrapper;
 import org.eclipse.yasson.defaultmapping.generics.model.ColoredCircle;
+import org.eclipse.yasson.defaultmapping.generics.model.ConstructorContainer;
 import org.eclipse.yasson.defaultmapping.generics.model.CyclicSubClass;
 import org.eclipse.yasson.defaultmapping.generics.model.FinalGenericWrapper;
 import org.eclipse.yasson.defaultmapping.generics.model.FinalMember;
@@ -49,6 +53,7 @@ import org.eclipse.yasson.defaultmapping.generics.model.MultipleBoundsContainer;
 import org.eclipse.yasson.defaultmapping.generics.model.MyCyclicGenericClass;
 import org.eclipse.yasson.defaultmapping.generics.model.PropagatedGenericClass;
 import org.eclipse.yasson.defaultmapping.generics.model.Shape;
+import org.eclipse.yasson.defaultmapping.generics.model.StaticCreatorContainer;
 import org.eclipse.yasson.defaultmapping.generics.model.WildCardClass;
 import org.eclipse.yasson.defaultmapping.generics.model.WildcardMultipleBoundsClass;
 import org.eclipse.yasson.serializers.model.Box;
@@ -473,6 +478,40 @@ public class GenericsTest {
         assertEquals(5, fromJson.get(0).field2.getArea());
         assertEquals(6, fromJson.get(0).field1);
         
+    }
+
+    @Test
+    public void genericConstructorCreator() {
+        final String expectedJson = "{\"value\":\"Test\"}";
+        final ConstructorContainer<String> container = new ConstructorContainer<>("Test");
+
+        assertEquals(expectedJson, defaultJsonb.toJson(container));
+        assertEquals(container, defaultJsonb.fromJson(expectedJson, ConstructorContainer.class));
+    }
+
+    @Test
+    public void genericStaticCreator() {
+        final String expectedJson = "{\"value\":\"static\"}";
+        final StaticCreatorContainer<String> container = StaticCreatorContainer.create("static");
+
+        assertEquals(expectedJson, defaultJsonb.toJson(container));
+        assertEquals(container, defaultJsonb.fromJson(expectedJson, StaticCreatorContainer.class));
+    }
+
+    @Test
+    public void wildcardCollectionContainer() {
+        final String expectedJson = "{\"collection\":{\"collection\":[{\"wrapped\":\"wrappedElement\"}]}}";
+        final CollectionContainer collectionContainer = new CollectionContainer();
+        final CollectionWrapper<CollectionElement<?>> collectionWrapper = new CollectionWrapper<>();
+        final CollectionElement<String> wildcardType = new CollectionElement<>();
+        wildcardType.setWrapped("wrappedElement");
+        final Collection<CollectionElement<?>> list = List.of(wildcardType);
+        collectionWrapper.setCollection(list);
+        collectionContainer.setCollection(collectionWrapper);
+
+        assertEquals(expectedJson, defaultJsonb.toJson(collectionContainer));
+        final CollectionContainer result = defaultJsonb.fromJson(expectedJson, CollectionContainer.class);
+        assertEquals(collectionContainer, result);
     }
     
     public interface FunctionalInterface<T> {

--- a/src/test/java/org/eclipse/yasson/defaultmapping/generics/model/CollectionContainer.java
+++ b/src/test/java/org/eclipse/yasson/defaultmapping/generics/model/CollectionContainer.java
@@ -1,0 +1,57 @@
+/*
+ * Copyright (c) 2025 Red Hat, Inc. and/or its affiliates.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0,
+ * or the Eclipse Distribution License v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+
+package org.eclipse.yasson.defaultmapping.generics.model;
+
+import java.util.Collection;
+import java.util.Objects;
+
+/**
+ * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
+ */
+public class CollectionContainer {
+
+    private CollectionWrapper<CollectionElement<?>> collection;
+
+    public CollectionWrapper<CollectionElement<?>> getCollection() {
+        return collection;
+    }
+
+    public void setCollection(final CollectionWrapper<CollectionElement<?>> collection) {
+        this.collection = collection;
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        if (obj == this) {
+            return true;
+        }
+        if (!(obj instanceof CollectionContainer)) {
+            return false;
+        }
+        final CollectionContainer other = (CollectionContainer) obj;
+        final Collection<CollectionElement<?>> thisCollection = collection.getCollection();
+        final Collection<CollectionElement<?>> otherCollection = other.collection.getCollection();
+        if (thisCollection == null && otherCollection == null) {
+            return true;
+        }
+        if (thisCollection == null || otherCollection == null) {
+            return false;
+        }
+        return thisCollection.containsAll(otherCollection) && otherCollection.containsAll(thisCollection);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(collection);
+    }
+}

--- a/src/test/java/org/eclipse/yasson/defaultmapping/generics/model/CollectionElement.java
+++ b/src/test/java/org/eclipse/yasson/defaultmapping/generics/model/CollectionElement.java
@@ -1,0 +1,48 @@
+/*
+ * Copyright (c) 2025 Red Hat, Inc. and/or its affiliates.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0,
+ * or the Eclipse Distribution License v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+
+package org.eclipse.yasson.defaultmapping.generics.model;
+
+import java.util.Objects;
+
+/**
+ * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
+ */
+public class CollectionElement<T> {
+
+    private T wrapped;
+
+    public T getWrapped() {
+        return wrapped;
+    }
+
+    public void setWrapped(T wrapped) {
+        this.wrapped = wrapped;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(wrapped);
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (!(obj instanceof CollectionElement)) {
+            return false;
+        }
+        final CollectionElement<?> other = (CollectionElement<?>) obj;
+        return Objects.equals(wrapped, other.wrapped);
+    }
+}

--- a/src/test/java/org/eclipse/yasson/defaultmapping/generics/model/ConstructorContainer.java
+++ b/src/test/java/org/eclipse/yasson/defaultmapping/generics/model/ConstructorContainer.java
@@ -1,0 +1,57 @@
+/*
+ * Copyright (c) 2025 Red Hat, Inc. and/or its affiliates.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0,
+ * or the Eclipse Distribution License v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+
+package org.eclipse.yasson.defaultmapping.generics.model;
+
+import java.util.Objects;
+
+import jakarta.json.bind.annotation.JsonbCreator;
+import jakarta.json.bind.annotation.JsonbProperty;
+
+/**
+ * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
+ */
+public class ConstructorContainer<T> {
+
+    private final T value;
+
+    @JsonbCreator
+    public ConstructorContainer(@JsonbProperty("value") final T value) {
+        this.value = value;
+    }
+
+    public T getValue() {
+        return value;
+    }
+
+    @Override
+    public String toString() {
+        return "ConstructorContainer[value=" + value + "]";
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(value);
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (!(obj instanceof ConstructorContainer)) {
+            return false;
+        }
+        final ConstructorContainer<?> other = (ConstructorContainer<?>) obj;
+        return Objects.equals(value, other.value);
+    }
+}

--- a/src/test/java/org/eclipse/yasson/defaultmapping/generics/model/StaticCreatorContainer.java
+++ b/src/test/java/org/eclipse/yasson/defaultmapping/generics/model/StaticCreatorContainer.java
@@ -1,0 +1,61 @@
+/*
+ * Copyright (c) 2025 Red Hat, Inc. and/or its affiliates.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0,
+ * or the Eclipse Distribution License v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+
+package org.eclipse.yasson.defaultmapping.generics.model;
+
+import java.util.Objects;
+
+import jakarta.json.bind.annotation.JsonbCreator;
+import jakarta.json.bind.annotation.JsonbProperty;
+
+/**
+ * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
+ */
+public class StaticCreatorContainer<T> {
+    private final T value;
+
+    private StaticCreatorContainer(T value) {
+        this.value = value;
+    }
+
+    @JsonbCreator
+    public static <T> StaticCreatorContainer<T> create(@JsonbProperty("value") final T value) {
+        return new StaticCreatorContainer<>(value);
+    }
+
+    public T getValue() {
+        return value;
+    }
+
+    @Override
+    public String toString() {
+        return "StaticCreatorContainer[value=" + value + "]";
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(value);
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (!(obj instanceof StaticCreatorContainer)) {
+            return false;
+        }
+        final StaticCreatorContainer<?> other = (StaticCreatorContainer<?>) obj;
+        return Objects.equals(value, other.value);
+    }
+
+}

